Fix changing the type of an existing Vault.

rjeffman opened the pull request:
Current implementation does not allow the change of an existingi Vault
type. To allow it, data is retrieved from the current vault, the vault
is modifiend, and then, data is stored again in the new vault.

Due to changing the process of modifying a vault, this change also
fixes the update of asymmetric vault keys. To change the key used,
the task must provide the old private key, used to retrieve data,
and the new public_key, used to store the data again. A new alias
was added to public_key (new_public_key) and public_key_file
(new_public_key_file) so that the playbook better express the
intention of the tak.

Vault tests have been updated to better test against the new update
process, and a new test file has bee added:

    tests/vault/test_vault_change_type.

t-woerner commented on tests/vault/test_vault_asymmetric.yml:
```diff
@@ -14,18 +14,101 @@
       name: asymvault
       vault_type: asymmetric
-      public_key: "{{ lookup('file', 'public.pem', rstrip=False) | b64encode }}"
+      public_key: "{{ lookup('file', 'old_public.pem', rstrip=True) | b64encode }}"
```
Please do not use the b64encode filter any more. We already removed it everywhere else.

rjeffman replied:
Done.

rjeffman commented:
The linter error is expected and in addressed in another PR (#471).

varunmylaraiah commented:
Thanks for the PR Rafael,
Now we can update the vault type from standard to symmetric and salt is also present.
Console output:
```
[root@ansible ~]# cat vault_standard_to_symmetric_vault.yaml
---
- name: Playbook to ensure, standard vault is changed to a symmetric type.
  hosts: ipaserver

  tasks:
  - ipavault:
      ipaadmin_password: <xxxxxx>
      name: std_to_symm
      vault_type: symmetric
      password: test_vault

Before:
[root@master ~]# ipa vault-show std_to_symm
  Vault name: std_to_symm
  Type: standard
  Owner users: admin
  Vault user: admin

After:
[root@master ~]# ipa vault-show std_to_symm
  Vault name: std_to_symm
  Type: symmetric
  Salt: yLmjVddfbi4domJjgpb2kRARYeO/EoILcKoRVTuI6DY=
  Owner users: admin
  Vault user: admin
```

varunmylaraiah commented:
Now we can update the symmetric vault to asymmetric vault but "Salt" is present in the asymmetric vault as well.
@rjeffman we need to fix this

Console-Output:
=============
```

[root@ansible ~]# cat vault_symmetric_to_asymmetric_vault.yaml
---
- name: Playbook to ensure symmetric vault changes to asymmetric type.
  hosts: ipaserver

  tasks:
  - ipavault:
      ipaadmin_password: <XXXXXXX>
      name: symm_to_asymm
      vault_type: asymmetric
      public_key: "{{ lookup('file', 'public.pem') | b64encode }}"

Before:
[root@master ~]# ipa vault-show symm_to_asymm
  Vault name: symm_to_asymm
  Type: symmetric
  Salt: e/azeysmtT5mJab2nTK48xW3uivdd6czRKFbCjwivI8=
  Owner users: admin
  Vault user: admin

After:
root@master ~]# ipa vault-show symm_to_asymm
  Vault name: symm_to_asymm
  Type: asymmetric
  Salt: e/azeysmtT5mJab2nTK48xW3uivdd6czRKFbCjwivI8=
  Public key: LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUF5cGZNeWdiaGtZRnRJVGdJanhONQpkNERoVXdoSnA1WjQ5N1dwTjJzUXpYOXNsVUgwSHJTNGtSMlJsVVgwbTNiQVhIOHhaYUp6TnhGcGF6eG8yU1NPCmRiOElmVGNRcFR0dmx2Uzd2azdvMlI1RXpKcmtMOENNeFhyWCtoK2FuVXFCZnI2WnhIZGZnb1pudndKNnh2WTQKbk5mWDRrejd2RnE3WUU0R21yZld0NDQ0Q2swSDU2dkdTeExuTC84OFZ1R2RhajIvVUVRQk15RWpKcUJQWkpaegpleEFmODMvUkFVdFNBN2dIU21Ba2JONnFNenNrL2dhYm93d1ppa1VaSUVCOERyRmlabFo1Mm41dTA2R1A3cUNqCjExVnhVZXZSZXhoclZva29WOFFSRDQvOGJxandnUUhxYjBHcnlsd3M4aDU4YzJ1MklYSFR4VXo5QVd2bmtQd1oKNFFJREFRQUIKLS0tLS1FTkQgUFVCTElDIEtFWS0tLS0t
  Owner users: admin
  Vault user: admin

```

varunmylaraiah commented:
Now we can update the asymmetric vault to a standard vault

Console-output
```

[root@ansible ~]# cat vault_asymmetric_to_standard_vault.yaml
---
- name: Playbook to ensure asymmetric vaults are changed to standard type.
  hosts: ipaserver

  tasks:
  - ipavault:
      ipaadmin_password: <xxxxxxx>
      name: asymm_to_std
      vault_type: standard

Before:
[root@master ~]# ipa vault-show asymm_to_std
  Vault name: asymm_to_std
  Type: asymmetric
  Public key: LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUF5cGZNeWdiaGtZRnRJVGdJanhONQpkNERoVXdoSnA1WjQ5N1dwTjJzUXpYOXNsVUgwSHJTNGtSMlJsVVgwbTNiQVhIOHhaYUp6TnhGcGF6eG8yU1NPCmRiOElmVGNRcFR0dmx2Uzd2azdvMlI1RXpKcmtMOENNeFhyWCtoK2FuVXFCZnI2WnhIZGZnb1pudndKNnh2WTQKbk5mWDRrejd2RnE3WUU0R21yZld0NDQ0Q2swSDU2dkdTeExuTC84OFZ1R2RhajIvVUVRQk15RWpKcUJQWkpaegpleEFmODMvUkFVdFNBN2dIU21Ba2JONnFNenNrL2dhYm93d1ppa1VaSUVCOERyRmlabFo1Mm41dTA2R1A3cUNqCjExVnhVZXZSZXhoclZva29WOFFSRDQvOGJxandnUUhxYjBHcnlsd3M4aDU4YzJ1MklYSFR4VXo5QVd2bmtQd1oKNFFJREFRQUIKLS0tLS1FTkQgUFVCTElDIEtFWS0tLS0t
  Owner users: admin
  Vault user: admin


After:
[root@master ~]# ipa vault-show asymm_to_std
  Vault name: asymm_to_std
  Type: standard
  Owner users: admin
  Vault user: admin
```

varunmylaraiah commented:
Now able to replace public-key-file to the public-key in asymmetric vault type

**Console-output:**

```
[root@ansible ~]# cat vault_asymmetric_vault_update_publickey_file_to_publickey.yaml
---
- name: Playbook to ensure the asymmetric vault is updated from public-key-file to public-key.
  hosts: ipaserver

  tasks:
  - ipavault:
      ipaadmin_password: <xxxxxxx>
      name: 76asyvault
      vault_type: asymmetric 
      vault_data: Test_Vault_Data
      public_key_file: "{{ ansible_env.HOME }}/public.pem"
  - ipavault:
      ipaadmin_password: <Xxxxxxxx>
      name: 76asyvault
      private_key_file: "{{ ansible_env.HOME }}/private.pem"
      public_key: |
        LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FR
        OEFNSUlCQ2dLQ0FRRUFsdDNDWk5CNzhNNmZmNXdOYm80OAp3ellWU1cxKzBOVFpXcUxtWUNiYk1m
        NmtmVHozazJvNEpnS0owUXY4VTZQK3c0SzkzVW5HUVN3WC9qRFh1TXNCCmxpN2JPSmh6MGQ5SlNw
        ZGo1NmJVNFNmQ0pQRmJlbUFiRGRtVkYzS1VtYUxOKzhBTEgzRy8vMk9KQlp0dzVndFkKbnN1bVIv
        ck1XRmg0eWZBSWhpemNKdlhLMlR5Qmxxc3R1N2VaZ2ErQ20wMDRIazhNNWdvbHBMQjE4aW1acW1Q
        YwovQ2d2b2dGajI5UHR5K3RISEtzK3J4Z2JpK2VwTWY2Z2xpV1JmbVdZc0JlSHJJaE41dHQ3d2Y2
        U0F2eHI5L2RkCi94cG9PcnNNUkhyRFd6bXJsSEN6cEJGbk82ajFuRWNvSk1mNFdBekxDSjVDOWIy
        dWpzR0JiTXZlYW1Oc2pMRjkKY1FJREFRQUIKLS0tLS1FTkQgUFVCTElDIEtFWS0tLS0tCg== 

[root@ansible ~]# ansible-playbook -i inventory/vault.hosts -vv vault_asymmetric_vault_update_publickey_file_to_publickey.yaml 
ansible-playbook 2.10.4
PLAYBOOK: vault_asymmetric_vault_update_publickey_file_to_publickey.yaml *********************************************************************************************************************************************************************
1 plays in vault_asymmetric_vault_update_publickey_file_to_publickey.yaml

PLAY [Playbook to ensure the asymmetric vault is updated from public-key-file to public-key.] ************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
task path: /root/vault_asymmetric_vault_update_publickey_file_to_publickey.yaml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [ipavault] ******************************************************************************************************************************************************************************************************************************
task path: /root/vault_asymmetric_vault_update_publickey_file_to_publickey.yaml:6
changed: [master.ipadomain.test] => {"changed": true}

TASK [ipavault] ******************************************************************************************************************************************************************************************************************************
task path: /root/vault_asymmetric_vault_update_publickey_file_to_publickey.yaml:12
changed: [master.ipadomain.test] => {"changed": true}
META: ran handlers
META: ran handlers

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
master.ipadomain.test      : ok=3    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```


**Try to retrieve data using the old private key file**
```

[root@ansible ~]# cat vault_asymmetric_vault_retrieve_data_using_old_private_key_file.yaml
---
- name: Playbook to ensure the data is not retrieved using old private-key-file.
  hosts: ipaserver

  tasks:
  - ipavault:
      ipaadmin_password: <xxxxx>
      name: 76asyvault
      private_key_file: "{{ ansible_env.HOME }}/private.pem"
      state: retrieved
      
      
[root@ansible ~]# ansible-playbook -i inventory/vault.hosts -vv vault_asymmetric_vault_retrieve_data_using_old_private_key_file.yaml 
ansible-playbook 2.10.4

PLAYBOOK: vault_asymmetric_vault_retrieve_data_using_old_private_key_file.yaml ***
1 plays in vault_asymmetric_vault_retrieve_data_using_old_private_key_file.yaml

PLAY [Playbook to ensure the data is not retrieved using old private-key-file.] ***

TASK [Gathering Facts] **********************************************************
task path: /root/vault_asymmetric_vault_retrieve_data_using_old_private_key_file.yaml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [ipavault] *****************************************************************
task path: /root/vault_asymmetric_vault_retrieve_data_using_old_private_key_file.yaml:6
fatal: [master.ipadomain.test]: FAILED! => {"changed": false, "msg": "vault_retrieve: 76asyvault: Invalid credentials"}

PLAY RECAP **********************************************************************
master.ipadomain.test      : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```


**Retrieve data using the new private key**

```
[root@ansible ~]# cat vault_asymmetric_vault_retrieve_data_using_new_private.yaml
---
- name: Playbook to ensure the data is retrieved from asymmetric vault with new private key.
  hosts: ipaserver

  tasks:
  - ipavault:
      ipaadmin_password: <xxxxxxxx>
      name: 76asyvault
      private_key: |
        LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFb3dJQkFBS0NBUUVBbHQzQ1pOQjc4
        TTZmZjV3TmJvNDh3ellWU1cxKzBOVFpXcUxtWUNiYk1mNmtmVHozCmsybzRKZ0tKMFF2OFU2UCt3
        NEs5M1VuR1FTd1gvakRYdU1zQmxpN2JPSmh6MGQ5SlNwZGo1NmJVNFNmQ0pQRmIKZW1BYkRkbVZG
        M0tVbWFMTis4QUxIM0cvLzJPSkJadHc1Z3RZbnN1bVIvck1XRmg0eWZBSWhpemNKdlhLMlR5Qgps
        cXN0dTdlWmdhK0NtMDA0SGs4TTVnb2xwTEIxOGltWnFtUGMvQ2d2b2dGajI5UHR5K3RISEtzK3J4
        Z2JpK2VwCk1mNmdsaVdSZm1XWXNCZUhySWhONXR0N3dmNlNBdnhyOS9kZC94cG9PcnNNUkhyRFd6
        bXJsSEN6cEJGbk82ajEKbkVjb0pNZjRXQXpMQ0o1QzliMnVqc0dCYk12ZWFtTnNqTEY5Y1FJREFR
        QUJBb0lCQUdmcUpWeTFjYmxEU1A5bgpxS1RPMmpHSHpnKyttTXdVNEFXYkVOMkh4aEZnbVVJZVBW
        VlBvNDhpem5Tc01jZ1d2SU5FU2F6amhPZ2REQVdzCjlObTNFc2UwVEpGbUNOaTFXTkZGQjY1bTds
        aUx2eUVkRWVwdnloSzNIUWVDaDdKdlBqdzZZaTd0ZGRiZVkySDIKV0VSTDZtV3VWa0ZjRVdhRXdl
        WXVMR0RuQTFqZGcyZFg4ckxSakhOTDRzMWtVZ1JqazRnemxsT2thR1pLNi9DRQpJMElsODI2NTJk
        WE13aHZYaU5qVVhWeWU3dDY3bGZWYmhkMzM2OUxwZ2JaVU54NG51UTRwSTRXcTRRYUF1RzFvCkE4
        RWtMejFrY2RTU3BVdSs4N0E5YVlDblErOEp4UXNyZ1gwTjJTa1hBcXpVNHRKYktVbjdnNmRTb2Z3
        d2tBK3IKb3Ryamw3RUNnWUVBeVFnMUxVVUxTb0pKRGt5c2d4ZWZGRmNBMGtPVmRaV3UxMTNCaWpD
        WHpKcjFQWmVEdnExUgo3WUp1WnE0bnpETmhVTGwzTkZab1orL3pjbWo4aERyMXlXNFVKM1czUGVj
        V3lTVjAveVFsZVN1SWZDUEQ4TzVqCkY2NjU4QkJnSjFOYUVxL0s0YnlJTnRyLzNWVUt3R0tndkJU
        WWlsL05kRzFjUmdZcEFLRCtFZDBDZ1lFQXdCNE8KdFE3aE42VS96VVlub1pkSUxHbEhHbVlONGFZ
        SnFsc3NVUVZjdURmaUg0ZndNbHJvdG1oR2JWdnRjdkJYKy9mZwpsYWFydUwvWUtaOGJER2NIVk9B
        T2pGWUlCaHlMWE9BTFlxNVB1OXhWYStUYVFMd3dMelA0VjNXL0VzOVI0L2F3Cjljd0F0d2QyYm05
        UDFDRnNIeExScmh5VHlDSm9GMG1EWm5UWlFxVUNnWUFROWNGT3ZXbHpjc1JwRVd6b2JtQzgKWFk2
        ekxvSjRkaFZYRkJEVThMNk53M0QzTGtZdUd2OHpjM0tPOVJoa2NOMC82elcwVTZuaFF1d2Q5ZDkv
        NUYyQwpEZWhqWDlvQ3ZmRTlMUXpPemFUZnNSMkZvNXJQUTcxV2JTWkFxNUVCSUZZVUI0S3RRck9Y
        TlNpcTZwYktua2VECk5wc0NLWXp5eXBzalNUZkNWdEs0ZlFLQmdGVjBTVGh5RU8ydWZHMnpXZWQv
        b2lQWEJFQ1N5blBuWmY4M3BNTzUKTGQvblE1YWc0YytVTGxVTWx0OEg5OXViV3BlQ1Q0eGhZWnJC
        QXNhb3RlQlRaNmh5TFNaeEVncWQ0Y09EQ0sxeApPWmduVG51NmhQWXNmd1pRQ0pQbWxRSXA0NVFS
        NCs3Qk9rUXZkcGR5dmNCdXlXZ2M5OVdKY3BhZjcwc0VjUVUrCnpuRGRBb0dCQUk1d1RFczY4M2li
        d2pwdVdvMlR5MzFMR1hYOS9YSG5QSHREOThjMXJsSGFWK2xzRTJRUHZueGsKQ1lXckl2ZnFSa2U1
        bmN1bTk4M1JUSkdFZys1ZzZsbTZjYVF4U3EwTi9FakVOSWFYWHJhS1A0UjA0eWRFM0YrMwpodmg4
        dEVyUHQrNDVpdTNXeTdWaEd6bDQzclhja05YVnFEYVZCRDgwU2V0eEtPazVsSW5tCi0tLS0tRU5E
        IFJTQSBQUklWQVRFIEtFWS0tLS0tCg==
      state: retrieved

[root@ansible ~]# ansible-playbook -i inventory/vault.hosts -vv vault_asymmetric_vault_retrieve_data_using_new_private.yaml
ansible-playbook 2.10.4

PLAYBOOK: vault_asymmetric_vault_retrieve_data_using_new_private.yaml ************************************************************************************************************************************************************************
1 plays in vault_asymmetric_vault_retrieve_data_using_new_private.yaml

PLAY [Playbook to ensure the data is retrieved from asymmetric vault with new private key.] **************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
task path: /root/vault_asymmetric_vault_retrieve_data_using_new_private.yaml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [ipavault] ******************************************************************************************************************************************************************************************************************************
task path: /root/vault_asymmetric_vault_retrieve_data_using_new_private.yaml:6
ok: [master.ipadomain.test] => {"changed": false, "vault": {"data": "Test_Vault_Data"}}
META: ran handlers
META: ran handlers

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
master.ipadomain.test      : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

rjeffman commented:
> Now we can update the symmetric vault to asymmetric vault but "Salt" is present in the asymmetric vault as well.
> @rjeffman we need to fix this

Hmm... it shouldn't... I'm looking into it.

rjeffman commented:
@varunmylaraiah I modified the code to cope with the attributes left when changing vault type. Tests were also modified to verify the changes.

Please, take a look at these changes.

varunmylaraiah commented:
Thanks, @rjeffman, 
Update the symmetric vault to asymmetric vault works fine.

```
Before:

[root@master ~]# ipa vault-show symm_to_asymm
  Vault name: symm_to_asymm
  Type: symmetric
  Salt: NN62zX73OJ8SaSn48+oTDRXUlaOsG+SrWXDSeNddlEQ=
  Owner users: admin
  Vault user: admin

After:

[root@master ~]# ipa vault-show symm_to_asymm
  Vault name: symm_to_asymm
  Type: asymmetric
  Public key: LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUFyK3JvOWMzOEs1STdRSkF3bGtJdApSRTloanBQbU9aVTV1akU3R3REZGdzZkc3S2IvTTFhZFlnTEhUelFTazZyV01WU0hXNGxzZDJPTlZ0TTBaZDlqCnNQckZ1Z3BVd3k3QUg2N2tJQnpNVHZHOG5CRS9NbFJ4WHJQNmdVdUFXU1ZyNGtjU1VHQmx2Q0FXSzk1alg4MXoKWVZ1a2Q1a00rNE5Nd1NSa1RrUi82NEdQeHlFdEEvRXdSVWdZRm1iZ2V6cjRXcEh0TVgvaDlQcWhQTkhDWjBPZgpPZXViTngyU3FZanR2cUd5MUxxeVFndTVreFRucGxnczVCTUtabjBoWFFMZDQwSkdlUElIUlBKaEtEOHBvSkNJCmZXV3haWXpKaEppc3ZVa2xwSHcyRXpqaHM1R1hyZ2h6LzBoYVFmNFBpd28yNlA5ZlZkbjIwc2ZDQm5icnNyRUIKVVFJREFRQUIKLS0tLS1FTkQgUFVCTElDIEtFWS0tLS0t
  Owner users: admin
  Vault user: admin

```

varunmylaraiah reviewed (approved):
All changes LGTM.

varunmylaraiah commented:
@t-woerner  We can merge this PR.

t-woerner reviewed (commented):
There are lookup calls in the new code with and without rstrip=True. Is that intentional?

rjeffman commented:
> There are lookup calls in the new code with and without rstrip=True. Is that intentional?

No, not intended. Removed all `rstrip: True`.

t-woerner reviewed (approved):
LGTM
